client/client_update_test.go

```diff
@@ -90,8 +101,341 @@ func testUpdateWithLocalCacheRemoteMissingMetadata(t *testing.T, forWrite bool)
+				} else {
+					require.True(t, bytes.Equal(expected, actual),
+						"%s: expected to update since only %s was unwritable", r, role)
+				}
+			}
+		}
+	}
+}
+
+type messUpMetadata func(role string) error
+
+// If a repo has corrupt metadata (in that the hash doesn't match the snapshot) or
+// missing metadata, an update will replace all of it
+func TestUpdateReplacesCorruptOrMissingMetadata(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode")
+	}
+	s, err := testutils.NewMetadataSwizzler("docker.com/notary")
+	require.NoError(t, err)
+
+	ts := readOnlyServer(t, s.MetadataCache, s.Roles)
+	defer ts.Close()
+
+	repo := newBlankRepo(t, ts.URL)
+	defer os.RemoveAll(repo.baseDir)
+
+	_, err = repo.Update(false) // ensure we have all metadata to start with
+	require.NoError(t, err)
+
+	// we want to swizzle the local cache, not the server, so create a new one
+	swizzler := testutils.MetadataSwizzler{
+		MetadataCache: repo.fileStore,
+		CryptoService: repo.CryptoService,
+		Roles:         s.Roles,
+	}
+
+	waysToMessUp := map[string]messUpMetadata{
+		"invalid JSON":     swizzler.SetInvalidJSON,
+		"missing metadata": swizzler.DeleteMetadata,
+	}
+	for _, role := range s.Roles {
+		for text, messItUp := range waysToMessUp {
+			for _, forWrite := range []bool{true, false} {
+				require.NoError(t, messItUp(role), "could not fuzz %s (%s)", role, text)
+				_, err := repo.Update(forWrite)
+				require.NoError(t, err)
+				for _, role := range s.Roles {
+					expected, err := s.MetadataCache.GetMeta(role, maxSize)
+					require.NoError(t, err, "problem getting expected metadata for %s", role)
+					actual, err := repo.fileStore.GetMeta(role, maxSize)
+					require.NoError(t, err, "problem getting repo metadata for %s", role)
+					require.True(t, bytes.Equal(expected, actual),
+						"%s for %s: expected to recover after update", text, role)
+				}
+			}
+		}
+	}
+}
+
+// If a repo has an invalid root (signed by wrong key, expired, invalid version, etc.),
+// the repo will just get the new root from the server, whether or not the update
+// is for writing (forced update)
+func TestUpdateWhenLocalRootRecoverablyCorrupt(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode")
+	}
+	s, err := testutils.NewMetadataSwizzler("docker.com/notary")
+	require.NoError(t, err)
+
+	ts := readOnlyServer(t, s.MetadataCache, s.Roles)
+	defer ts.Close()
+
+	repo := newBlankRepo(t, ts.URL)
+	defer os.RemoveAll(repo.baseDir)
+
+	_, err = repo.Update(false) // ensure we have all metadata to start with
+	require.NoError(t, err)
+
+	// we want to swizzle the local cache, not the server, so create a new one
+	swizzler := testutils.MetadataSwizzler{
+		MetadataCache: repo.fileStore,
+		CryptoService: s.CryptoService,
+		Roles:         s.Roles,
+	}
+
+	waysToMessUp := map[string]messUpMetadata{
+		// TODO: If invalid threshold fails because it's an invalid role, then this
```

Comment:
After some discussion with @cyli, I agree that we can eliminate those tests -- we should catch invalid signatures, but if a user has signed his own local root metadata to be invalid I'm not sure what we can do other than have him remove his local copy right?